### PR TITLE
GEODE-3539: Add missing test coverage for 'list regions' and 'describe region' commands

### PR DESCRIPTION
**Pull request by PurelyApplied**
* Mutually exclusive options --group and --member checked in interceptor instead of failing server-side
* Command logic flattened to stream.
* Testing of 'list region' and 'describe region' separated into distinct DUnit and Integration tests
* Removed @Flaky annotation from 'describe region' DUnit test.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/ListRegionDUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+  public static GfshShellConnectionRule gfshShellConnectionRule = new GfshShellConnectionRule();
+
+  @BeforeClass
+  public static void setupSystem() throws Exception {
+    final Properties locatorProps = createProperties("Locator", "G3");
+    MemberVM locator = lsRule.startLocatorVM(0, locatorProps);
+
+    final Properties managerProps = createProperties("Manager", "G1");
+    managerProps.setProperty(LOCATORS, "localhost[" + locator.getPort() + "]");
+    MemberVM manager = lsRule.startServerVM(1, managerProps, locator.getPort());
```
Comment (jdeppe-pivotal) [inline geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/ListRegionDUnitTest.java]: This isn't really a manager, is it?

File: geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/ListRegionDUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  private static final String LOCALREGIONONMANAGER = "LocalRegionOnManager";
+
+  @ClassRule
+  public static LocatorServerStartupRule lsRule = new LocatorServerStartupRule();
+
+  @ClassRule
+  public static GfshShellConnectionRule gfshShellConnectionRule = new GfshShellConnectionRule();
+
+  @BeforeClass
+  public static void setupSystem() throws Exception {
```
Comment (jdeppe-pivotal) [inline geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/ListRegionDUnitTest.java]: There's a ton of setup in here that doesn't seem to be relevant to what is actually being tested. Can this be trimmed down to the minimum to get the tests to pass?

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/ListRegionCommand.java [hunk 3]
```diff
@@ -110,4 +108,17 @@ public Result listRegion(
     }
     return result;
   }
+
+  public static class ListRegionInterceptor extends AbstractCliAroundInterceptor {
```
Comment (jinmeiliao) [inline geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/ListRegionCommand.java]: the call of getMember or findMember inside the command will throw an exception if both are set, so we don't really need to do the check here. I admit that this brings the validation to the client side, but it's a bit of a redundant code. I think we can afford to make that check on the server side if it makes the code a little cleaner.

File: geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/DescribeRegionIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  public GfshShellConnectionRule gfsh = new GfshShellConnectionRule();
+
+  @Test
+  public void commandFailsWhenNotConnected() throws Exception {
+    gfsh.executeAndAssertThat("describe region").statusIsError()
+        .containsOutput("was found but is not currently available");
+  }
+
+  @Test
+  public void commandFailsWithoutNameOption() throws Exception {
```
Comment (jinmeiliao) [inline geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/DescribeRegionIntegrationTest.java]: just a minor point: invalid command that failed parser takes a long time to run in the dunit/integration test. Better put them in a unit test and uses GfshParserRule to test them.

Comment (PurelyApplied) [reply geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/DescribeRegionIntegrationTest.java]: Would that be preferable to just lowering the timeout on `HeadlessGfsh` for commands that should be executed quickly?  I see that the timeout is otherwise only set at instantiation of the `HeadlessGfsh` instance.

File: geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/ListRegionDUnitTest.java (hunk 2 above)
Comment (PurelyApplied) [reply geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/ListRegionDUnitTest.java]: Good catch.  I pulled this from a test that was testing both `list regions` and `describe region` and didn't look closely enough at it.

File: geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/DescribeRegionIntegrationTest.java (hunk 4 above)
Comment (jinmeiliao) [reply geode-core/src/test/java/org/apache/geode/management/internal/cli/commands/DescribeRegionIntegrationTest.java]: yeah, that would be a good solution too.

File: geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/ListRegionCommand.java (hunk 3 above)
Comment (PurelyApplied) [reply geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/ListRegionCommand.java]: I think cleaner is better, too, but throwing inside the command server-side feels a little too close to the "exception as program flow" antipattern.  Exceptions should be exceptional, right?

There's also the issue that such an exception then gets wrapped a couple of times, to result in the output
```
Could not process command due to error. Error occurred while fetching list of regions. : Please provide either "member" or "group" option.
```
This could be improved by a better message in the root exception, but still would have two confusing layers of output for something that is ultimately a User error, not a server-error.

Comment (jinmeiliao) [reply geode-core/src/main/java/org/apache/geode/management/internal/cli/commands/ListRegionCommand.java]: You probably did not rebase to the current develop, I just made a change in CommandExecutor to process UserErrorException differently than general exception. That would probably ease your concern. 

The getMemberXXX call would have to throw an exception anyway to catch the cases where some commands do not do that validation in their interceptors.

**Comment by PurelyApplied** [comment]
Rebased and updated per requests, precheckin running.

Changes to `GfshCommandRule` and `HeadlessGfsh` copied directly from PR #1065.  Verify that no strange conflict exists before rebasing / merging.
